Basic/DyanmicPrefabs/Assets/Scripts/03_Server Authoritative Try Spawn Synchronously/SynchronousDynamicPrefabSpawn.cs

```diff
@@ -0,0 +1,145 @@
+            TrySpawnSynchronously();
+        }
+        
+        async void TrySpawnSynchronously()
+        {
+            var randomPrefab = m_DynamicPrefabReferences[Random.Range(0, m_DynamicPrefabReferences.Count)];
+            await TrySpawnDynamicPrefabSynchronously(randomPrefab.AssetGUID, Random.insideUnitCircle * 5, Quaternion.identity);
+        }
+
+        /// <summary>
+        /// This call attempts to spawn a prefab by it's addressable guid - it ensures that all the clients have loaded the prefab before spawning it,
+        /// and if the clients fail to acknowledge that they've loaded a prefab - the spawn will fail.
+        /// </summary>
+        /// <param name="guid"></param>
+        /// <returns></returns>
+        async Task<(bool Success, NetworkObject Obj)> TrySpawnDynamicPrefabSynchronously(string guid, Vector3 position, Quaternion rotation)
+        {
+            if (IsServer)
+            {
+                var assetGuid = new AddressableGUID()
+                {
+                    Value = guid
+                };
+                
+                if (DynamicPrefabLoadingUtilities.IsPrefabLoadedOnAllClients(assetGuid))
+                {
+                    Debug.Log("Prefab is already loaded by all peers, we can spawn it immediately");
+                    var obj = await Spawn(assetGuid);
+                    return (true, obj);
+                }
+                
+                m_SynchronousSpawnAckCount = 0;
+                m_SynchronousSpawnTimeoutTimer = 0;
+                
+                Debug.Log("Loading dynamic prefab on the clients...");
+                LoadAddressableClientRpc(assetGuid);
+                //load the prefab on the server, so that any late-joiner will need to load that prefab also
+                await DynamicPrefabLoadingUtilities.LoadDynamicPrefab(assetGuid, m_ArtificialDelayMilliseconds);
+                var requiredAcknowledgementsCount = IsHost ? m_NetworkManager.ConnectedClients.Count - 1 : 
+                    m_NetworkManager.ConnectedClients.Count;
+                
+                while (m_SynchronousSpawnTimeoutTimer < m_SpawnTimeoutInSeconds)
+                {
+                    if (m_SynchronousSpawnAckCount >= requiredAcknowledgementsCount)
+                    {
+                        Debug.Log($"All clients have loaded the prefab in {m_SynchronousSpawnTimeoutTimer} seconds, spawning the prefab on the server...");
+                        var obj = await Spawn(assetGuid);
+                        return (true, obj);
+                    }
+                    
+                    m_SynchronousSpawnTimeoutTimer += Time.deltaTime;
+                    await Task.Yield();
+                }
+                
+                Debug.LogError("Failed to spawn dynamic prefab - timeout");
+                return (false, null);
+            }
+
+            return (false, null);
+
+            async Task<NetworkObject> Spawn(AddressableGUID assetGuid)
```

Comment:
name is a bit misleading. Could be "TryLoadAndSpawn"?
Cause I was looking higher up at the Spawn when the object is already loaded and I was wondering "wait, why is it async when we know the addressable is already loaded".
Could there be a version without an Addressable load?

Reply:
Hmm now looking at it again, at both points where Spawn(..) is invoked, the Addressable is already loaded on the server. I feel like it can be spawned directly. Let me look into that.